cli/crates/federated-dev/src/dev/composer.rs

```diff
@@ -0,0 +1,170 @@
+
+pub(crate) use self::subgraph::Subgraph;
+
+use super::{
+    bus::{ComposeBus, ComposeMessage, ComposeSchema, IntrospectSchema, RemoveSubgraph},
+    refresher::RefreshMessage,
+};
+use crate::error::Error;
+use async_graphql_parser::parse_schema;
+use grafbase_graphql_introspection::introspect;
+use graphql_composition::{compose, Subgraphs};
+use std::collections::BTreeMap;
+
+pub(crate) struct Composer {
+    bus: ComposeBus,
+    graphs: BTreeMap<String, Subgraph>,
+}
+
+impl Composer {
+    pub(crate) fn new(bus: ComposeBus) -> Self {
+        Self {
+            bus,
+            graphs: BTreeMap::default(),
+        }
+    }
+
+    pub(crate) async fn handler(mut self) -> Result<(), crate::Error> {
+        loop {
+            match self.bus.recv().await {
+                Some(ComposeMessage::Introspect(message)) => {
+                    self.handle_introspect(message).await?;
+                }
+                Some(ComposeMessage::Compose(message)) => {
+                    self.handle_compose(message).await?;
+                }
+                Some(ComposeMessage::RemoveSubgraph(message)) => {
+                    self.handle_remove_subgraph(message).await?;
+                }
+                Some(ComposeMessage::Recompose) => {
+                    self.handle_recompose().await?;
+                }
+                Some(ComposeMessage::InitializeRefresh) => {
+                    self.handle_init_refresh().await?;
+                }
+                None => break,
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_introspect(&mut self, message: IntrospectSchema) -> Result<(), crate::Error> {
+        let (name, url, headers, responder) = message.into_parts();
+
+        let headers = headers
+            .iter()
+            .map(|header| (header.key(), header.value()))
+            .collect::<Vec<_>>();
+
+        let result = introspect(url.as_str(), headers.as_slice())
+            .await
+            .and_then(|sdl| parse_schema(sdl).map_err(|error| error.to_string()));
+
+        match result {
+            Ok(schema) => {
+                responder
+                    .send(Ok(schema))
+                    .map_err(|_| Error::internal("oneshot channel dead"))?;
+            }
+            Err(error) => {
+                let error = Error::introspection(error.to_string());
+
+                responder
+                    .send(Err(error))
+                    .map_err(|_| Error::internal("oneshot channel dead"))?;
+
+                self.bus.send_composer(RemoveSubgraph::new(&name)).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn ingest_subgraphs(&self, add_new: Option<(&str, &Subgraph)>) -> Subgraphs {
+        let mut subgraphs = Subgraphs::default();
+
+        for (name, subgraph) in &self.graphs {
+            if add_new.is_some_and(|(added_name, _)| added_name == name) {
+                continue;
+            }
+
+            subgraphs.ingest(subgraph.schema(), name);
```

Comment:
I just changed the signature of (I think) this function btw - sorry for the breakage.  Probably want to pull in `main` and add `subgraph.url()` (or whatever it is) on the end